feat: put request id in details dictionary

rvm4 opened the pull request:
* closes #388, #412, #415.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
[Preview](https://s3.amazonaws.com/pr-preview/w3c/browser-payment-api/request_id_construct.html) | [Diff](https://s3.amazonaws.com/pr-preview/w3c/browser-payment-api/3c3a100...113fb60.html)

rvm4 committed:
put request id in details dictionary

rsolomakhin commented:
Looks like you're passing in a `paymentRequestID` to the constructor, but remove its accessors and don't mention where this identifier is going.

If you expect the merchant website to read this identifier, please expose it through an accessor in `PaymentResponse` interface. Depending on your use-case, you may want to expose the identifier through an accessor in `PaymentRequest` interface as well.

If you expect the payment app to read the identifier, please add a note somewhere that it should be sent to the payment app.

marcoscaceres commented on index.html:
```diff
@@ -301,6 +300,7 @@
           </p>
           <pre class="example" title="The 'details' argument">
             const details = {
+              paymentRequestID: "1234567890123",
```
s/paymentRequestID/paymentRequestId

marcoscaceres commented on index.html:
```diff
@@ -1014,6 +1014,7 @@
       </h2>
       <pre class="idl">
         dictionary PaymentDetails {
+          DOMString paymentRequestID;
```
s/paymentRequestID/paymentRequestId/g

marcoscaceres commented on index.html:
```diff
@@ -1033,6 +1034,17 @@
       </p>
       <dl>
         <dt>
+          <dfn>paymentRequestID</dfn>
```
s/paymentRequestID/paymentRequestId

Commit:
Lowercase d in Id, as per platform conventions

marcoscaceres commented on index.html:
```diff
@@ -1033,6 +1034,17 @@
+        </dt>
+        <dd>
+          <a>paymentRequestId</a> is a freeform identifier for this
+          payment request. This field is intended to hold a unique
```
Re: "intended", as it reads, it could be confused that the browser is supposed to use it for error recovery. I would remove the second sentence.

marcoscaceres commented on index.html:
```diff
@@ -1033,6 +1034,17 @@
+          payment request. This field is intended to hold a unique
+          identifier for the purposes of recovering from failures and/or as a
+          foreign key into other systems. If a <a>paymentRequestId</a>
+          is not provided by the client then a UUID will be generated and stored
```
The third sentence should be a non-normative note. Also, there is no requirement that this shall be a UUID in the payment request constructor.

marcoscaceres commented on index.html:
```diff
@@ -204,7 +204,6 @@
           Promise&lt;void&gt; abort();
           Promise&lt;boolean&gt; canMakePayment();
 
-          readonly attribute DOMString? paymentRequestId;
```
Did you delete this by accident from the interface?

marcoscaceres replied:
Also, the constructor algorithm doesn't ever set this to null - so maybe it should not be nullable. It's also unclear what happens if paymentRequestId is set to just "" (the empty string). 

Should setting it to `null` be treated as a special case (i.e., treat null as undefined instead of being converted to the word "null")?

domenic replied:
We should not treat `null` as a special case. It is no more special than `false` or `0` (all of them are "things that are vaguely negative, but are not the default-triggering special value of `undefined`").

marcoscaceres commented on index.html:
```diff
@@ -460,6 +457,9 @@
           </li>
           <li>Let <var>serializedModifierData</var> be an empty list.
+          <li>If <var>details.paymentRequestId</var> was not provided during
+          construction, populate the field with an [[RFC4122]] UUID.
```
Oh, I see. I don't think we have consensus to make this a hard requirement. Also, avoid modifying the incoming data, create a new variable instead.

marcoscaceres replied:
Drafting some alternative text...

marcoscaceres replied:
Sent some text for the comments above.

marcoscaceres commented on index.html:
```diff
@@ -333,9 +333,6 @@
           The <a>PaymentRequest</a> constructor MUST act as follows:
         </p>
         <ol data-link-for="PaymentDetails" class="algorithm">
-          <li>If a <code>paymentRequestId</code> was not provided during
```
Here, I think we want this instead:

```HTML
          <li>Establish the request's id:
            <ol>
              <li>Let <var>requestId</var> be the empty string.
              </li>
              <li>If <var>details.id</var> is present, set <var>
                requestId</var> to the value of
                <var>details.paymentRequestId</var>. Otherwise, generate a
                string to uniquely identify this payment request and set
                <var>requestId</var> to that value. It is RECOMMENDED that the
                generated string be a <abbr title=
                "Universally Unique Identifier">UUID</abbr> as per
                [[!RFC4122]].
              </li>
            </ol>
          </li>
```

Then, after the object is actually constructed:

```HTML
          <li>Set <var>request</var>.<a>paymentRequestId</a> to
          <var>requestId</var>.
```

marcoscaceres commented:
Fixed the nits via a commit. Let me know if you agree with the changes I proposed. I can add them in.

domenic commented:
I realize this is a bit late in the process, so feel free to ignore me, but is there a reason it's `paymentRequest.paymentRequestId` instead of `paymentRequest.id`?

marcoscaceres commented:
I was thinking the same. It's not too late to change it.

marcoscaceres commented:
(no one shipped this, so we can change still)

Commit:
Add paymentRequestId back in as "id"

marcoscaceres commented:
@domenic, so, in practice you would end up with (not ideal, IMO):

``` js
const request = new PaymentRequest(x, {paymentRequestId: "123121"},z);
const response = await request.show();
// stuff happens... 
if (request.id === response.paymentRequestId ) {

}
```
What might be best is just call it `requestId` across all three: 

```JS
const request = new PaymentRequest(x, {requestId: "123312"}, z);
const response = await request.show();
// stuff happens... 
if (request.requestId === response.requestId ) {

}
```

WDYT?

domenic commented:
`requestId` could be reasonable, although I was thinking something more like

```js
const request = new PaymentRequest(x, {id: "123121"},z);
const response = await request.show();
// stuff happens... 
if (request.id === response.requestId) {

}
```

marcoscaceres commented:
@domenic, yeah, that works well actually... it's only the response that needs the qualifier.

Commit:
Feat: Add PaymentResponse.requestId

Commit:
fix: id is not nullable 

The payment request Id is either provided by the developer or automagically generated by the UA - hence not nullable right now.

Commit:
rename paymentRequestId to just id

Commit:
fix: use id instead of paymentRequestId

marcoscaceres committed:
update text to match discussion

marcoscaceres committed:
Set the response id attribute

marcoscaceres commented:
I've updated the PR to reflect discussion above:

 * `PaymentRequest` gains "id" attribute.
 * `PaymentResponse` gains "requestId" attribute.
 * `PaymentDetailsBase` gains "id" member.
 * PaymentRequest constructor algo gains "Establish the request's id:" sub-algo.  
 * UUID is RECOMMENDED.
 * PaymentResponse's .requestId gets set on construction.

marcoscaceres commented:
Question: do we want an internal slot for the [id] for when it's copied over from Request to Response? Can it be forged if the attribute value is just copied over? I guess having the internal slot is nicer - but I don't have an strong opinion.

marcoscaceres committed:
use [[requestId]] internal slot instead

marcoscaceres commented:
so, decided internal slots are fun - so added one. PR now reflects that.

rsolomakhin commented on index.html:
```diff
@@ -830,6 +850,14 @@
+              <dfn>[[\requestId]]</dfn>
+            </td>
+            <td>
+              A string, which can be empty, that identifies the a payment request.
```
s/the a/the/

rsolomakhin reviewed (approved):
LGTM

domenic commented on index.html:
```diff
@@ -344,6 +342,17 @@
           </li>
+          <li>Establish the request's id:
+            <ol>
+              <li>If <var>details.id</var> is present, set <a>[[\requestId]]</a>
```
Elsewhere (e.g. for details.shippingOptions) we've used the approach of normalizing the details.whatever, instead of creating a separate slot. I think that'd work best here too, so no separate internal slot for the ID. A separate slot leads to potential confusion where algorithms could accidentally reach into both places and get different values.

domenic commented on index.html:
```diff
@@ -591,6 +602,15 @@
+        </h2>
+        <p>
+          When getting, the <a>id</a> attribute reflects the value of
+          the payment request's <a>[[\requestId]]</a> internal slot.
```
"Reflects" is a term of art that isn't quite appropriate here. You want "When getting, returns this PaymentRequest's [[requestId]] internal slot" (or, per the above advice, "this PaymentRequest's [[details]].id")

domenic commented on index.html:
```diff
@@ -1033,6 +1062,17 @@
+        <dd>
+          <a>id</a> is a free-form identifier for this payment request.
+          <p class="note">
+            If a <a>id</a> member is not present, then the UA will generate a
```
s/a/an; maybe link "user agent" instead of using "UA"

domenic commented on index.html:
```diff
@@ -1033,6 +1062,17 @@
+          <p class="note">
+            If a <a>id</a> member is not present, then the UA will generate a
+            unique identifier for the payment request during <a data-lt=
+            "PaymentRequest.PaymentRequest()">construction</a>.
```
This is a broken link according to ReSpec.

marcoscaceres replied:
Yep, I've got a patch to fix that in ReSpec. I'll try to land that soon.

domenic reviewed (changes requested):
Processing model looks good, just a few nits and the issue of where to store it.

marcoscaceres committed:
Editorial: improve custom details.id usage example

marcoscaceres committed:
editorial: make it clear we are adding an .id member

marcoscaceres commented:
Rebased

domenic commented on index.html:
```diff
@@ -583,6 +595,16 @@
+        </h2>
+        <p>
+          When getting, the <a>id</a> attribute returns this
+          <a>PaymentRequest</a> <a>[[\details]]</a>.<a data-lt=
```
Nit: missing `'s` after PaymentRequest

domenic commented on index.html:
```diff
@@ -1035,7 +1057,9 @@
       </h2>
       <pre class="idl">
-        dictionary PaymentDetailsBase {
+        dictionary PaymentDetails {
```
Rebasing seems to have gone bad here.

marcoscaceres replied:
Oh, we'll spotted. How odd :(

domenic commented on index.html:
```diff
@@ -1047,6 +1071,29 @@
+          </p>
+        </dd>
+        <dt>
+          <dfn>total</dfn>
```
and here

marcoscaceres commented:
@domenic, sorry, can you recheck this one (request.id)? I'm not if everything was addressed or not.

domenic reviewed (changes requested):
Bad rebase but overall looking good to me.

marcoscaceres committed:
fix rebase issue

marcoscaceres commented:
Ok, I think I fixed the rebase issues.

domenic commented:
Oh, I think this also needs handling in updateWith()? Unless it should move to PaymentDetailsInit...

marcoscaceres commented:
I'm not sure it makes sense in `.updateWith()`, so PaymentDetailsInit would be the right place.

marcoscaceres committed:
transplant id to PaymentDetailsInit

marcoscaceres commented on index.html:
```diff
@@ -1131,6 +1154,17 @@
+          <p class="note">
+            If an <a>id</a> member is not present, then the <a>user agent</a>
+            will generate a unique identifier for the payment request during
+            <a data-lt="PaymentRequest.PaymentRequest()">construction</a>.
```
ReSpec will link this soon...
